uvm/hier/usr/lib/python3/dist-packages/uvm/login_tools.py

```diff
@@ -29,6 +40,76 @@ def get_auth_uri():
         return 'https://auth.untangle.com/v1/CheckTokenAccess'
 
 
+class Totp:
+    """
+    Manaage TOTP for time-based OTP
+    """
+    cmd_file_path = "/usr/share/untangle/conf/cmd_totp.conf"
+
+    @classmethod
+    def enabled(cls, req):
+        """
+        Detect if TOTP is enabled for this login
+
+        :returns True if TOTP is enabled, False otherwise
+        """
+        if os.path.isfile(Totp.cmd_file_path) is False:
+            # No seed file
+            return False
+
+        if Totp.is_local(req):
+            # Not enabled for a local console session
+            return False
+
+        if pyopt_imported is False:
+            # If we're unable to import the pyopt library,
+            # we're not enabled but we want to be very explicit about this case!
+            apache.log_error("Error!  Unable to import pyotp")
+            return False
+
+        return True
+
+    @classmethod
+    def is_local(cls, req):
+        """
+        Detect if client is using the local console.
+        :param req          Request object.
+        :returns            True if client is using local console, False othewise.
+        """
+        (client_addr, client_port) = req.useragent_addr
+
+        return client_addr == "127.0.0.1" or client_addr == "::1"
+
```

Comment:
The way I read 'validate' it seems like we have a single seed for TOTP for all Administrative and Reports accounts? Is that what we want?